### PR TITLE
P0: Document Ren'Py engine adapter contract

### PR DESCRIPTION
**Pull request by hu-border-collie**
Refs #265
Refs #285

## Summary

- record the current Ren'Py discover, scan, build, relocate, validate, render, check, apply, progress, and RAG backfill call chains against the current checkout
- separate engine-neutral ownership from Ren'Py-specific parsing, locator, validation, and rendering semantics
- define the P1/P2 engine adapter protocol and versioned occurrence, opaque locator, validation result, writeback plan, and localization mode contracts
- define candidate inventory, coverage report/review, freshness, digest composition, review policy, provenance, statuses, and reason-code registries
- define the first-stage boundaries for sync, Batch, revision, keyword, Project Analysis, Final Review, progress, and RAG
- add direct characterization tests for scanning/source identity/speaker, duplicate block identity, v2 relocation, and multi-span string rendering

## Scope guard

This PR is P0 design and characterization only. It does not add a production adapter, enable coverage gates, change manifest/config schemas, alter CLI or GUI behavior, migrate revision/keyword/writeback paths, or implement any part of P1-P6.

The contract keeps `TranslationUnit` as the only translation core model. An occurrence is a version/location envelope around the existing unit; identity v2 and manifest v1/v2 remain unchanged. Adapters return declarative plans and never receive arbitrary file-write authority. Project/source validation, the latest safe check, path containment, source re-read, transaction recovery, and atomic writeback remain in the common layer.

## Validation

- `python -m unittest tests.test_engine_adapter_p0_characterization tests.test_translator_runtime tests.test_identity_v2 tests.test_atomic_io tests.test_batch_golden_corpus` — 157 passed, 1 skipped
- `python -B tests/run_cli_tests.py -q` — 784 passed, 1 skipped
- local Markdown link check — 32 files, all local targets exist
- `git diff --cached --check` — passed before commit

## Summary by CodeRabbit

- **Documentation**
  - Added a detailed P0 design document for Ren’Py engine adapter integration, covering call-chain behavior, data contracts, compatibility boundaries, readiness criteria, and future staging.
  - Updated planning documentation with links to the new design and its coverage scope.

- **Tests**
  - Added characterization tests covering translation-unit identity, repeated blocks, line-drift stability, relocation behavior, unresolved items, and replacement rendering with quote escaping.

### CODE REVIEW

**Review by hu-border-collie — COMMENTED** [review]
## 审查结论：建议通过（COMMENT；无法对自己的 PR 投 Approve）

**范围：** 设计文档 + characterization 测试（无生产 adapter）  
**CI：** 全绿  
**本地：** `tests.test_engine_adapter_p0_characterization` 4/4 通过

### 总评

这是一份扎实的 #265 / #285 P0 交付。范围边界清楚，调用链审计大体贴合当前代码，安全归属（adapter 只返回声明式 plan、公共层保留 check/apply）正确，Coverage 不能自证完整的合同也写到位。P1 切口（先统一 sync/Batch translation extraction，revision/writeback 放 P2）务实。

| 维度 | 评价 |
|---|---|
| 范围纪律 | 很好 |
| 调用链审计 | 强 |
| Adapter / 安全合同 | 强 |
| Coverage / review 合同 | 强 |
| Characterization 基线 | 好；扫描器分叉尚未钉死 |
| 作为 P0 可合入性 | 可以 |

### 做得好的地方

1. **硬边界明确**——第 1 节与 PR 正文有效防止顺手实现 P1。
2. **调用链审计贴代码**——三种翻译扫描 + 一种剧情结构扫描的诊断准确。
3. **安全归属正确**——`--force` 只放宽“已 apply”防重，不绕过 safe check / 源校验 / `block`。
4. **身份模型拆分清楚**——`TranslationUnit` 仍是唯一核心；occurrence 为信封；`occurrence_id` / `content_fingerprint` / `lineage_id` 分离。
5. **覆盖审计不能自证**——inventory → audit → 独立 review → confirmation；Agent/人工 provenance 与 stale 规则清晰。
6. **P1 切入口务实**——共享 extraction 先行，避免最高风险的 dual-write 迁移。
7. **Characterization 测试真实**——固定 speaker/source marker/重复 block、行漂移 identity、v2 relocation、多 span 逆序渲染与 quote 转义。

### 建议处理（详见行内评论）

| 优先级 | 主题 |
|---|---|
| 中等（本 PR 最好补） | locator / writeback plan 的 0/1-based 行号与半开列区间 |
| 中等（高价值可选项） | 同名双 `collect_translation_entries_from_lines` + 扫描器分叉尚未被测试钉住 |
| 低 | Project Analysis 入口模块归属 |
| 低 | `validate_translation` bool+message → reason code 映射草表 |
| 低 | writeback“谁负责渲染 fragment”一句定论 |
| nit | §5.2 id 示例标明“示意” |

### 与 #285 验收对照

调用链、归属、无写权限、schema/freshness/digest/policy、工作流边界、characterization 基线、manifest/identity/check→apply 兼容、独立 review/stale、无用户可见行为变化——均满足。扫描面 characterization 偏窄，但不阻塞 P0 设计合入。

**结论：** 建议通过。因 GitHub 不允许对自己的 PR 投 Approve，此处用 COMMENT 提交完整审查与行内建议；行内评论为合入前后应补清的设计约定，避免 P1/P2 实现期踩坑。

File: docs/plans/engine_adapter_contract.md [hunk 1]
```diff
@@ -0,0 +1,763 @@
...
+  "locator": {
+    "file_rel_path": "chapter1.rpy",
+    "translate_block": "chapter_1_abcd",
+    "block_occurrence": 1,
+    "ordinal": 42,
+    "line_hint": 120,
+    "start_col_hint": 8,
+    "end_col_hint": 31,
+    "source_marker_kind": "comment"
+  }
```
Comment (hu-border-collie) [inline docs/plans/engine_adapter_contract.md:329]: ### 中等：行号/列号索引约定未写清

示例里 `line_hint: 120`，而 §5.5 writeback plan 示例是 `line: 119`。当前生产代码约定是：

- `line`：0-based
- `line_number`：1-based
- `start` / `end`：该行上的列偏移，半开区间 `[start, end)`（tokenizer token 边界）

这里看起来像“hint 用 1-based、plan 用 0-based”，但正文没有显式声明。P2 很容易把 locator hint 和 plan span 混用。

**建议：** 在 §5.3 明确每个字段的 0/1-based 与是否半开；并写明公共层只校验 plan 上的 span，不把 locator hint 当业务坐标。

File: docs/plans/engine_adapter_contract.md [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+      "target_root": "localization_catalog",
+      "target_rel_path": "chapter1.rpy",
+      "expected_file_sha256": "sha256...",
+      "line": 119,
+      "start_col": 8,
+      "end_col": 31,
+      "expected_fragment_sha256": "sha256...",
+      "expected_text_digest": "sha256...",
+      "replacement_fragment": "\"译文\"",
+      "validation_digest": "sha256..."
```
Comment (hu-border-collie) [inline docs/plans/engine_adapter_contract.md:418]: ### 中等（续）：WritebackPlan 坐标语义

`line` / `start_col` / `end_col` 需要与现有 `WritebackAction` / relocation 输出对齐并写死约定：

1. `line` 是否 0-based（建议与现状一致）
2. `start_col`/`end_col` 是否半开
3. 与 locator 的 `line_hint` / `*_col_hint` 是否同一索引体系，还是 hint 故意用 1-based 人类可读坐标

另外：`replacement_fragment` 已是带引号的最终片段（如 `\"译文\"`）。

**建议在 §5.5 加一句定论：** adapter 负责 quote/escape 并产出最终 fragment；公共层只做 span/hash 校验与拼接，不再二次走 `render_replacement_lines` / `quote_with`。

File: docs/plans/engine_adapter_contract.md [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+
+因此“Ren'Py scan”目前至少有三种不同目的的扫描：
+
+1. `collect_tasks_with_progress()`：待翻译 task 与已译计数；
+2. `scan_all_translation_units()`：identity v2 到 live span 的重定位表；
+3. `collect_translation_entries_from_lines()` /
+   `collect_repair_entries_from_lines()`：revision、keyword、repair、RAG 的
+   source/translation pair。
+
+Project Analysis 另有第四种结构扫描，不能和翻译候选扫描互相证明完整。
```
Comment (hu-border-collie) [inline docs/plans/engine_adapter_contract.md:58]: ### 中等：三种扫描器分叉 + 同名双实现

“三种不同目的的扫描”诊断正确，且 §9.2 风险 2 已点名规则并非完全同一实现。补充一点代码事实：

- `translator_runtime.collect_translation_entries_from_lines(lines)` —— sync RAG 等路径，无 identity v2 绑定
- `gemini_translate_batch.collect_translation_entries_from_lines(lines, file_rel_path='')` —— revision/RAG 等，内部走 `build_identity_v2_by_span`

同名不同实现，签名与语义都不完全一致。§2.3/§2.4 若只写一个函数名，读者容易以为是一条链。

**建议：** 在此节或 §10 写明“同名双实现是现状债务”；P1 等价性目前只对 `collect_tasks_with_progress` + `scan_all_translation_units` 主张（与文末 characterization 一致）。

File: docs/plans/engine_adapter_contract.md [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+[`tests/test_engine_adapter_p0_characterization.py`](../../tests/test_engine_adapter_p0_characterization.py)
+直接固定：
+
+- source marker、speaker、translated count 与重复 translate block 的扫描输出；
+- identity v2 在行漂移后的稳定性和 duplicate block occurrence 区分；
+- `relocate_v2_chunk_items()` 按 live identity 更新 span，并报告 unresolved item；
+- `render_replacement_lines()` 的多 replacement 逆序应用、quote escaping 和输入不变。
+
+这些测试描述当前行为，不授权 P0 修改生产实现。P1 adapter 必须先在这些结果上做到
+等价，再单独增加 candidate/error reporting。
```
Comment (hu-border-collie) [inline docs/plans/engine_adapter_contract.md]: ### 中等（测试覆盖）：扫描器分叉尚未被钉住

当前 characterization 很好地固定了：

- pending/relocation 扫描（source marker、speaker、重复 block、行漂移）
- v2 relocation + unresolved
- multi-span reverse-order render + quote escaping

但未覆盖 §9.2 风险 2 提到的 revision/repair 路径：

- batch `collect_translation_entries_from_lines`
- runtime 同名 collector（对比）
- `collect_repair_entries_from_lines`
- voice / keyword-argument 跳过、source marker 不成对等

对纯 P0 设计合入不是硬挡，但是合入后 / P1 开工前最值得补的锁。

**建议（高价值可选项）：** 加一条 fixture，用同一组 lines 对比 runtime vs batch collector 输出差异，或至少在本节声明“分叉未 pin，P1 不得仅包装其中一个就宣称统一”。

File: docs/plans/engine_adapter_contract.md [hunk 5]
```diff
@@ -0,0 +1,763 @@
...
+清理旧记录。因此 RAG 是 adapter 之后的派生 consumer，不是 occurrence、lineage
+或译文历史的事实来源。
+
+### 2.4 revision、keyword、Project Analysis 与 Final Review
+
+| 路径 | 当前扫描 / 消费方式 | P0 判断 |
+|---|---|---|
+| revision | `collect_revision_file_jobs() -> collect_translation_entries_from_lines()`；后者再用 `build_identity_v2_by_span() -> scan_all_translation_units(mode=revision)` 绑定 identity v2。preview/apply 继续走 source validation、Ren'Py renderer 和原子写回。 | 已译 occurrence 提取与 translation pending 扫描相邻但不同；P1 不抢先迁移，P2 与 relocation/validation/writeback plan 一起收敛。 |
+| keyword | `collect_keyword_file_jobs() -> collect_repair_entries_from_lines()`，同时合并 source/translation pair 与 `legacy.collect_tasks()`。输出 glossary candidates，不写 `.rpy`。 | P1 保持原状；以后消费 candidate inventory 的“可分析文本投影”，不能把关键词任务等同翻译覆盖。 |
+| Project Analysis | `build_structure_drafts() -> discover_script_files()` 扫 source `.rpy` 且跳过 `tl/`，再由 `build_route_graph()` 解析 label/jump/call/menu；content fingerprint 与 lineage 存在自己的 store。 | 保留独立剧情结构 parser。后续只把 fresh coverage digest/review status 作为 publish upstream gate；不把全部 UI 文本注入分析 prompt。 |
```
Comment (hu-border-collie) [inline docs/plans/engine_adapter_contract.md]: ### 低：模块归属略偏

`build_structure_drafts()` 实际在 `project_analysis_generate.py`；`discover_script_files` / `build_route_graph` 在 `project_analysis_routes.py`。

行为描述正确，入口模块差一跳。

**建议：** 写成 `project_analysis_generate.build_structure_drafts() -> discover_script_files() / build_route_graph()`，避免实现时找错入口。

File: docs/plans/engine_adapter_contract.md [hunk 6]
```diff
@@ -0,0 +1,763 @@
...
+  `common.target_language.missing`、`common.locator.unresolved`；
+- Ren'Py：`renpy.placeholder.missing`、`renpy.placeholder.added`、
+  `renpy.tag.changed`、`renpy.field.changed`、`renpy.percent_token.changed`、
+  `renpy.string_literal.unrenderable`；
+- writeback preflight：`writeback.project_mismatch`、
+  `writeback.source_snapshot_mismatch`、`writeback.span_mismatch`、
+  `writeback.path_escape`、`writeback.overlap`、`writeback.plan_stale`。
+
+现有 Batch `CHECK_WARN_REASON_CODES` / `CHECK_BLOCK_REASON_CODES` 保持原样；上述
+adapter reason codes 在 P2 需要显式映射，不能通过错误字符串推断。
```
Comment (hu-border-collie) [inline docs/plans/engine_adapter_contract.md:391]: ### 低：补一张 current → v1 映射草表

现状 `translator_runtime.validate_translation` 返回 `(bool, message)`，没有 reason code。这里已正确要求 P2 对 Batch `CHECK_*_REASON_CODES` 做显式映射，但 bool+message → `pass|warn|block` + 稳定 code 仍缺草表。

**建议加简表（可草稿）：**

| 当前 message 模式 | 建议 code | 默认 status |
|---|---|---|
| Empty translation | `common.translation.empty` | block |
| Preserved terms missing | `common.preserve_term.missing` | block |
| Ren'Py placeholders/tags changed | `renpy.placeholder.*` / tag / percent | block |
| No Chinese characters | `common.target_language.missing` | block |

避免 P2 自行发明不兼容严重度。

File: docs/plans/engine_adapter_contract.md [hunk 7]
```diff
@@ -0,0 +1,763 @@
...
+  "content_fingerprint": "sha256...",
+  "candidate_id": "cand1:sha256...",
+  "locator": {
+    "engine": "renpy",
+    "locator_schema_version": 1,
+    "locator": {}
+  },
+  "translation_unit": {
+    "core_schema_version": 2,
+    "id": "script.rpy:chapter_1:42:abcd1234",
```
Comment (hu-border-collie) [inline docs/plans/engine_adapter_contract.md:283]: ### nit：示例 ID 标明“示意”

真实 identity v2 形态是 `file:block[#occ]:ordinal:sha1[:8]`。这个示例作 shape 可以，建议加“示意，非真实 hash”，减少后续文档/测试 cargo-cult。

File: tests/test_engine_adapter_p0_characterization.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
+# -*- coding: utf-8 -*-
+"""P0 characterization tests for the current Ren'Py parsing boundary.
+
+These tests intentionally exercise existing helpers directly. P0 does not
+introduce an adapter implementation; P1 must preserve these outputs while it
+makes parse failures and unsupported candidates observable.
+"""
```
Comment (hu-border-collie) [inline tests/test_engine_adapter_p0_characterization.py:7]: ### 肯定 + 覆盖边界

这组 characterization 质量很好：直接钉现有 helper、明确“P0 不改生产实现、P1 先等价再加可观测错误”。

本地已跑通 4/4。

覆盖边界提醒（与设计文档同行评论一致）：当前未 pin revision/repair 双 collector 分叉。P0 可接受；P1 宣称“扫描统一”前建议至少补一条对比测试。
